### PR TITLE
rlp: add ArrayVec impls

### PR DESCRIPTION
**Pull request by vorot93**
Useful for limited-size lists.

### CODE REVIEW

File: rlp/src/impls.rs [hunk 1]
```diff
@@ -79,6 +79,44 @@ impl Decodable for Vec<u8> {
 	}
 }
 
+#[cfg(feature = "arrayvec")]
```
Comment (dvdplm) [inline rlp/src/impls.rs:82]: I think you missed adding this feature to `Cargo.toml`.

**Review by dvdplm — CHANGES_REQUESTED** [review]
A few tests would be good: normal, zero-length, beyond capacity.

**Comment by dvdplm** [comment]
I don't want to be naysayer but this is a slippery slope here; `ArrayVec` yes but `SmallVec`, `TinyVec`, `ThinVec` no? It would be good to have a concrete use-case and a proper motivation for why `ArrayVec` will be useful long-term.

File: rlp/src/impls.rs (hunk 1 above)
Comment (vorot93) [reply rlp/src/impls.rs:82]: Optional dependencies are features in themselves.

**Comment by vorot93** [comment]
My concrete use case is for `Neighbours` discv4 packet:

https://github.com/ethereum/devp2p/blob/master/discv4.md#neighbors-packet-0x04

It is a given that the reply is a limited-size vec of up to 16 elements. Would be nice to embed this property in the type system, but for that I need RLP to support _any_ limited-size vec implementation. I suppose `ArrayVec` is the most popular out there, so I just went with that.

File: rlp/src/impls.rs (hunk 1 above)
Comment (dvdplm) [reply rlp/src/impls.rs:82]: TIL! 👍 

(I'd still want all features in the `Cargo.toml`, preferably documented, for discoverability reasons)

**Comment by dvdplm** [comment]
Thank you for the motivation, it's much clearer now. I'm still hesitant. Even if it's an optional dependency, it still needs maintenance. The future of `arrayvec` is a bit unclear, as the author puts it ["Const generics is arrayvec 2.0"](https://github.com/bluss/arrayvec/issues/48#issuecomment-433688621), so I think this would be useful mostly until const generics is ready?

**Comment by ordian** [comment]
@vorot93 is this PR still relevant now that we have `Bytes` impl?

**Comment by vorot93** [comment]
@ordian Yes! This PR is about limited collections of items, not byte arrays.

**Comment by ordian** [comment]
I see, I thought the reason was to avoid allocations which, it seems, you were able to achieve with `bytes`.
I share the same opinion as @dvdplm and hesitant to merge this as min const generics are targeting stabilization in february.
For now you can use a newtype wrapper for `ArrayVec`.

**Pull request closed without merging**